---
title: "scRNASeq / snSeqRNA"
author: "cellbusters"
date: "2024-06-25"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(Seurat)
library(Matrix)
library(sctransform)
```

Read in the single cell data.

```{r}
path2matrix <- function(path_csv_gz, data_name) {
  input_df           = read.csv(path_csv_gz)
  rownames(input_df) = input_df$X
  input_df           = input_df[, -1]
  input_matrix       = as.sparse(input_df)
  input_matrix
  # obj = CreateSeuratObject(counts=input_matrix, project=data_name)
}
```


```{r}
cell_exon   = path2matrix('../data/GSE123454_cells_exon_counts.csv.gz',   "cell_exon")
cell_intron = path2matrix('../data/GSE123454_cells_intron_counts.csv.gz', "cell_intron")

nuc_exon   = path2matrix('../data/GSE123454_nuclei_exon_counts.csv.gz',   "nuc_exon")
nuc_intron = path2matrix('../data/GSE123454_nuclei_intron_counts.csv.gz', "nuc_exon")
```

Verify that the gene IDs are in the same order in both data sets.

```{r}
stopifnot(rownames(cell_exon) == rownames(cell_intron))
stopifnot(rownames(cell_exon) == rownames(nuc_exon))
stopifnot(rownames(cell_exon) == rownames(nuc_intron))
```

Combining data & create Seurat objects.

```{r}
data_nuc  = nuc_exon  + nuc_intron
data_cell = cell_exon + cell_intron

obj_nuc  = CreateSeuratObject(counts = data_nuc,  project = "data_nuc")
obj_cell = CreateSeuratObject(counts = data_cell, project = "data_cell")
```

Create a merged Seurat object.

```{r}
# Merge samples
merged.obj <- merge(obj_nuc, y = c(obj_cell),
              add.cell.ids = c("nuclei", "cells"), project = "cellbusters")

merged.obj$orig.ident <- sapply(X = strsplit(colnames(merged.obj), split = "_"), 
                                FUN = "[", 1)
```

Add mitochoncrial and ribosomal gene content.

```{r}
merged.obj[["percent.mt"]] <- PercentageFeatureSet(merged.obj, pattern = "^[Mm][Tt]-")

ribo_genes = rownames(cell_exon)[grep('^Rp[sl]', rownames(cell_exon))]
ribo_genes = ribo_genes[!grepl('ps[0-9]+$', ribo_genes)]
merged.obj[["percent.ribo"]] <- PercentageFeatureSet(merged.obj, features = ribo_genes)
```

What is the distribution of intron and exon counts in each data set?

```{r}
df = data.frame(gene_id     = rownames(cell_exon),
                cell_exon   = rowSums(cell_exon,   na.rm = TRUE),
                cell_intron = rowSums(cell_intron, na.rm = TRUE),
                nuc_exon    = rowSums(nuc_exon,    na.rm = TRUE),
                nuc_intron  = rowSums(nuc_intron,  na.rm = TRUE))
```

```{r}
df %>%
  pivot_longer(cols = cell_exon:nuc_intron) %>%
  separate(name, into = c('Assay', 'Exon/Intron')) %>%
  mutate(value = value + 1) %>%
  ggplot(aes(value)) +
    geom_histogram(bins = 100) +
    scale_x_log10() +
    facet_grid(Assay ~ `Exon/Intron`) +
    lims(y = c(0, 1000)) +
    labs(title = 'Distribution of Exon and Intron Reads')
```

Plot the proportion of intron reads for each gene in sc & sn.

```{r}
df = data.frame(sc_prop = rowSums(cell_intron) / rowSums(cell_intron + cell_exon),
                sn_prop = rowSums(nuc_intron)  / rowSums(nuc_intron  + nuc_exon))

df %>%
  filter(!is.nan(sc_prop) & ! is.nan(sn_prop)) %>%
  pivot_longer(cols = sc_prop:sn_prop) %>%
  mutate(name = if_else(name == 'sc_prop', 'Single Cell', 'Single Nuclear')) %>%
  ggplot(aes(value)) +
    geom_histogram(bins = 100) +
    facet_wrap(~name, ncol = 1, scales = 'free_y') + 
    labs(title = 'Intronic Proportions in Each Gene')
```

Plot the proportion of intron reads for each cell in sc & sn.

```{r}
df = data.frame(sc_prop = colSums(cell_intron) / colSums(cell_intron + cell_exon),
                sn_prop = colSums(nuc_intron)  / colSums(nuc_intron + nuc_exon))

df %>%
  filter(!is.nan(sc_prop) & ! is.nan(sn_prop)) %>%
  pivot_longer(cols = sc_prop:sn_prop) %>%
  mutate(name = if_else(name == 'sc_prop', 'Single Cell', 'Single Nuclear')) %>%
  ggplot(aes(value)) +
    geom_histogram(bins = 100) +
    facet_wrap(~name, ncol = 1, scales = 'free_y') + 
    labs(title = 'Intronic Proportions in Each Cell')
```

```{r}
tibble(x = merged.obj$orig.ident, 
       y = merged.obj$percent.ribo) %>%
  ggplot(aes(x, y, fill=x))
```


```{r}
ggplot(tibble(x=merged.obj$orig.ident, y=merged.obj$nFeature_RNA),aes(x, y, fill=x)) +
  geom_boxplot() +
  guides(fill=F) +
labs(title = "Genes per Cell", x = "Method", y = "nFeatures_RNA") +
  theme(text = element_text(size = 20))
```

```{r}
ggplot(tibble(x=merged.obj$orig.ident, y=merged.obj$nCount_RNA),aes(x, y, fill=x)) +
  geom_boxplot() +
  guides(fill=F) +
  labs(title = "Counts per Cell", x = "Method", y = "nCount_RNA") +
  theme(text = element_text(size = 20))
```

```{r}
ggplot(tibble(x=merged.obj$orig.ident, y=merged.obj$percent.mt),aes(x, y, fill=x)) +
  geom_boxplot() +
  guides(fill=F) +
  labs(title = "Mitochondrial RNA Content", x = "Method", y = "% Mitochondrial RNA") +
  theme(text = element_text(size = 20))
```

```{r}
ggplot(tibble(x = merged.obj$orig.ident, y = merged.obj$percent.ribo), aes(x, y, fill = x)) +
  geom_boxplot() +
  guides(fill = F) +
  labs(title = "Ribosomal RNA Content", x = "Method", y = "% Ribosomal RNA") +
  theme(text = element_text(size = 20))
```

Normalize, scale data, run PCA, and clustering.


```{r}

```


